docs/RFCS/20220602_fine_grained_cpu_attribution.md

```diff
@@ -0,0 +1,615 @@
+    0's got  9.98% of total time
+    1's got  9.53% of total time
+    2's got  9.22% of total time
+    3's got 10.42% of total time
+    4's got  9.84% of total time
+    5's got 10.43% of total time
+    6's got 10.50% of total time
+    7's got 10.21% of total time
+    8's got 10.03% of total time
+    9's got  9.86% of total time
+
+=== RUN   TestProportionalGoroutines           # want incrementing multipliers
+    0's got  1.87% of total time (1.000000x)
+    1's got  3.60% of total time (1.931999x)
+    2's got  5.41% of total time (2.899312x)
+    3's got  7.21% of total time (3.864451x)
+    4's got  9.11% of total time (4.880925x)
+    5's got 10.94% of total time (5.864723x)
+    6's got 12.77% of total time (6.842004x)
+    7's got 14.34% of total time (7.685840x)
+    8's got 16.58% of total time (8.885060x)
+    9's got 18.18% of total time (9.741030x)
+```
+
+TODO: Evaluate inaccuracy given time spent descheduled by the OS is also
+counted here. The Go scheduler tries its best to hog the processors it’s been
+allotted, and given we’re running goroutines much higher than the number of
+processors, the observed inaccuracy across all goroutines (only the ones pinned
+to a descheduled OS thread is overcounted) should be diminishingly small.
+
+## Alternatives 
+
+There are a few options to get CPU attribution with varying degrees of
+accuracy, granularity and convenience. It’s possible we incorporate one or more
+of the listed alternatives in the future.
+
+### Sampling profiler
+
+We can get coarse-grained, non-instantaneous CPU attribution using pprof and
+[profiler labels](https://rakyll.org/profiler-labels/). A sketch of what this
+would look like was [prototyped
+here](https://github.com/cockroachdb/cockroach/pull/60508), and has roughly the
+following idea.
+- We'd make use of profiler labels at points of interest (code paths where
+  requests to a single range/tenant feed through for e.g.)
+- Go automatically propagates profiler labels across goroutine boundaries with
+  shared contexts. gRPC also propagates labels across RPC boundaries.
+- We'd periodically capture profiling data and use it to compute the %es of CPU
+  time attributable to specific scopes.
+
+Comparison:
+
+- Does not necessitate runtime changes.
+- Provides coarse-grained attribution, which is likely sufficient for things
+  like the allocator or the tenant rate limiter, but harder to apply to
+  per-statement/request accounting.
+    - Sampling rates have a maximum frequency of 100hz with implications on
+      fidelity.
+    - Sampling-backed signals are accurate over a larger number of samples
+      (i.e. non-instantaneous), which is less-than-usable when we care about
+      capturing CPU data for executions in the tail (think statements that
+      don't take that long, don't happen frequently).
+- CPU profiling with labels is allocation inefficient and has a higher
+  performance overhead (though bounded to <1%; TODO: more numbers). Profiles
```

Comment:
Are you going to revise the RFC after for this TODO?

Reply:
Not immediately, there's some discussion in the internal gdoc about the kind of experimental setup we'd need here to make this more apples-to-apples and it's work I'm happy deferring until needed. I'll keep this TODO to indicate it would still be worth doing for completeness.

Reply:
sounds good.